// @figliolia/chalk-animation/dist/mjs/Animation.js
import { FrameRate } from "./FrameRate";
import { Decor } from "./Decor";
export class Animation extends FrameRate {
    lines;
    text;
    stopped = false;
    initialized = false;
    effect;
    static LINE_BREAKS = /\r\n|\r|\n/;
    constructor(effect, str, delay = 0, speed = 0) {
        super(delay, speed);
        this.effect = effect;
        this.text = str.split(Animation.LINE_BREAKS);
        this.lines = this.text.length;
    }
    render() {
        if (!this.initialized) {
            Animation.LOG("\n".repeat(this.lines - 1));
            this.initialized = true;
        }
        Animation.LOG(this.nextFrame());
        this.schedule(() => {
            if (!this.stopped) {
                this.render();
            }
        });
    }
    nextFrame() {
        this.increment();
        return ("\u001B[" +
            `${this.lines}` +
            "F\u001B[G\u001B[2K" +
            this.text.map(str => this.effect(str, this.current)).join("\n"));
    }
    replace = Decor.chainable(this, (str) => {
        this.text = str.split(Animation.LINE_BREAKS);
        this.lines = this.text.length;
    });
    stop = Decor.chainable(this, () => {
        this.clearFrames();
        this.stopped = true;
    });
    start = Decor.chainable(this, () => {
        this.clearFrames();
        this.stopped = false;
        this.render();
    });
}
